source/common/common/fancy_logger.h

```diff
@@ -0,0 +1,111 @@
+   */
+  void setDefaultFancyLevelFormat(spdlog::level::level_enum level, std::string format)
+      ABSL_LOCKS_EXCLUDED(fancy_log_lock_);
+
+private:
+  /**
+   * Initializes sink for the initialization of loggers, needed only in benchmark test.
+   */
+  void initSink();
+
+  /**
+   * Creates a logger given key and log level, and add it to map.
+   * Key is the log component name, e.g. file name now.
+   */
+  spdlog::logger* createLogger(std::string key, int level = -1)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(fancy_log_lock_);
+
+  /**
+   * Lock for the following map (not for the corresponding loggers).
+   */
+  absl::Mutex fancy_log_lock_;
+
+  /**
+   * Map that stores <key, logger> pairs, key can be the file name.
+   */
+  FancyMapPtr fancy_log_map_ ABSL_GUARDED_BY(fancy_log_lock_) = std::make_shared<FancyMap>();
+};
+
+FancyContext& getFancyContext();
+
+#define FANCY_KEY std::string(__FILE__)
+
+/**
+ * Macro for fancy logger.
+ * Uses a global map to store logger and take use of thread-safe spdlog::logger.
+ * The local pointer is used to avoid another load() when logging. Here we use
+ * spdlog::logger* as atomic<shared_ptr> is a C++20 feature.
+ */
+#define FANCY_LOG(LEVEL, ...)                                                                      \
+  do {                                                                                             \
+    static std::atomic<spdlog::logger*> flogger{0};                                                \
+    spdlog::logger* local_flogger = flogger.load(std::memory_order_relaxed);                       \
+    if (!local_flogger) {                                                                          \
+      getFancyContext().initFancyLogger(FANCY_KEY, flogger);                                       \
+      local_flogger = flogger.load(std::memory_order_relaxed);                                     \
+    }                                                                                              \
+    local_flogger->log(spdlog::source_loc{__FILE__, __LINE__, __func__},                           \
+                       ENVOY_SPDLOG_LEVEL(LEVEL), __VA_ARGS__);                                    \
+  } while (0)
+
+/**
+ * Convenient macro for connection log.
+ */
+#define FANCY_CONN_LOG(LEVEL, FORMAT, CONNECTION, ...)                                             \
+  FANCY_LOG(LEVEL, "[C{}] " FORMAT, (CONNECTION).id(), ##__VA_ARGS__)
+
+/**
+ * Convenient macro for stream log.
+ */
+#define FANCY_STREAM_LOG(LEVEL, FORMAT, STREAM, ...)                                               \
+  FANCY_LOG(LEVEL, "[C{}][S{}] " FORMAT, (STREAM).connection() ? (STREAM).connection()->id() : 0,  \
+            (STREAM).streamId(), ##__VA_ARGS__)
+
+/**
+ * Convenient macro for log flush.
+ */
+#define FANCY_FLUSH_LOG()                                                                          \
+  { getFancyContext().getFancyLogEntry(FANCY_KEY)->flush(); }
```

Comment:
nit: wrap with `do {} while(0)` or remove wrapping `{}`

Reply:
Addressed and will show in next commit.